enzyme/Enzyme/MLIR/Analysis/DataFlowActivityAnalysis.cpp

```diff
@@ -0,0 +1,976 @@
+          if (valueState->getValue().isActiveVal()) {
+            auto *ptrAliasClass = getOrCreateFor<AliasClassLattice>(op, value);
+            forEachAliasedAlloc(ptrAliasClass, [&](DistinctAttr alloc) {
+              result |= before->setActiveOut(alloc);
+            });
+          }
+        }
+      }
+      if (isa<MemoryEffects::Write>(effect.getEffect())) {
+        auto *ptrAliasClass = getOrCreateFor<AliasClassLattice>(op, value);
+        std::optional<Value> stored = getStored(op);
+        std::optional<Value> copySource = getCopySource(op);
+        forEachAliasedAlloc(ptrAliasClass, [&](DistinctAttr alloc) {
+          if (stored.has_value() && after.activeDataFlowsOut(alloc)) {
+            if (isa<FloatType, ComplexType>(stored->getType())) {
+              auto *valueState = getOrCreate<BackwardValueActivity>(*stored);
+              propagateIfChanged(
+                  valueState, valueState->meet(ValueActivity::getActiveVal()));
+            }
+          } else if (copySource.has_value() &&
+                     after.activeDataFlowsOut(alloc)) {
+            auto *srcAliasClass =
+                getOrCreateFor<AliasClassLattice>(op, *copySource);
+            forEachAliasedAlloc(srcAliasClass, [&](DistinctAttr srcAlloc) {
+              before->setActiveOut(srcAlloc);
+            });
+          } else if (auto linalgOp = dyn_cast<linalg::LinalgOp>(op)) {
+            if (after.activeDataFlowsOut(alloc)) {
+              for (OpOperand &dpsInit : linalgOp.getDpsInitsMutable()) {
+                if (dpsInit.get() == value) {
+                  int64_t resultIndex =
+                      dpsInit.getOperandNumber() - linalgOp.getNumDpsInputs();
+                  Value yieldOperand =
+                      linalgOp.getBlock()->getTerminator()->getOperand(
+                          resultIndex);
+                  auto *valueState =
+                      getOrCreate<BackwardValueActivity>(yieldOperand);
+                  propagateIfChanged(
+                      valueState,
+                      valueState->meet(ValueActivity::getActiveVal()));
+                }
+              }
+            }
+          }
+        });
+      }
+    }
+    propagateIfChanged(before, result);
+  }
+
+  void visitCallControlFlowTransfer(CallOpInterface call,
+                                    CallControlFlowAction action,
+                                    const BackwardMemoryActivity &after,
+                                    BackwardMemoryActivity *before) override {
+    meet(before, after);
+  }
+
+  void setToExitState(BackwardMemoryActivity *lattice) override {}
+
+private:
+  FunctionOpInterface parentOp;
+  ArrayRef<enzyme::Activity> argumentActivity;
+};
+
+void traverseCallGraph(FunctionOpInterface root,
+                       SymbolTableCollection *symbolTable,
+                       function_ref<void(FunctionOpInterface)> processFunc) {
+  std::deque<FunctionOpInterface> frontier{root};
```

Comment:
Nit: using `SmallVector` is likely much more memory-efficient than `deque` (or `queue`). Though the traversal is not a pure BFS, it's still valid.